README.md
# mflux-ai-python
Open source code for the mflux-ai python package


HISTORY.rst
=======
History
=======

v0.7.0 (TBA)
-------------------

* Implement :code:`mflux_ai.get_best_run()`, which returns the best run in a model group defined in MFlux.ai.

v0.6.0 (2019-12-13)
-------------------

* Move functions from mflux_ai.mflux_ai to mflux_ai.core. The API stays backwards-compatible for now, but warnings are shown if the old API is used. This change was applied to get cleaner and more consistent import statements.

v0.5.3 (2019-09-23)
-------------------

* Improve the performance and the support for special characters in object names in :code:`mflux_ai.get_dataset` by unpickling in memory instead of using a temporary file on disk.
* Don't expose non-public variables and imports on the top-level package

v0.5.2 (2019-09-20)
-------------------

* Improve the performance of :code:`mflux_ai.put_dataset` by pickling in memory instead of using a temporary file on disk.

v0.5.1 (2019-09-12)
-------------------

* Add support for MLflow authentication
* Improve the performance of :code:`mflux_ai.put_dataset`
* Correctly reset the MinIO client when :code:`init` completes successfully
* Specify the desired API version and let the user know if an upgrade is needed

v0.4.0 (2019-09-01)
-------------------

* Mark :code:`mflux_ai.set_env_vars()` as deprecated. Use :code:`mflux_ai.init()` instead.
* Remove support for Python 3.4
* Add support for secure MinIO connections

v0.3.0 (2019-08-16)
-------------------

* Add a function :code:`init` that will eventually replace :code:`set_env_vars`
* Check if the provided project token is valid.

v0.2.1 (2019-08-16)
-------------------

* Set licence to Apache License 2.0
* Transition from pre-alpha to alpha.

v0.2.0 (2019-08-14)
-------------------

* Add convenience functions for storing and retrieving datasets

v0.1.1 (2019-08-14)
-------------------

* First release on PyPI. Has support for setting environment variables for MLflow based on an MFlux.ai project token.


tests/test_core.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for the `mflux_ai` package."""
import os
import warnings

import responses


import mlflow
import mflux_ai
from mflux_ai.core import SERVER_HOST


@responses.activate
def test_mflux_ai_init():
    """Test the init function."""
    content = {
        "minio_secret_key": "minio_secret",
        "minio_access_key": "minio_access",
        "minio_server": "http://192.198.0.1:9000",
        "mlflow_server": "http://192.198.0.1:5000",
    }

    responses.add(
        responses.Response(
            method="GET", url=SERVER_HOST + "/api/env_vars/", json=content, status=200
        )
    )

    mflux_ai.init("thisshouldbevalidtoken")
    assert os.environ.get("MLFLOW_TRACKING_URI") == content["mlflow_server"]
    assert os.environ.get("MLFLOW_S3_ENDPOINT_URL") == content["minio_server"]
    assert os.environ.get("AWS_SECRET_ACCESS_KEY") == content["minio_secret_key"]
    assert os.environ.get("AWS_ACCESS_KEY_ID") == content["minio_access_key"]
    assert os.environ.get("MLFLOW_TRACKING_TOKEN") == "thisshouldbevalidtoken"


@responses.activate
def test_mflux_ai_deprecated_set_env_vars():
    """Test the deprecated set_env_vars function."""
    content = {
        "minio_secret_key": "minio_secret",
        "minio_access_key": "minio_access",
        "minio_server": "http://192.198.0.1:9000",
        "mlflow_server": "http://192.198.0.1:5000",
    }

    responses.add(
        responses.Response(
            method="GET", url=SERVER_HOST + "/api/env_vars/", json=content, status=200
        )
    )

    with warnings.catch_warnings(record=True) as w:
        # Cause all warnings to always be triggered.
        warnings.simplefilter("always")
        mflux_ai.set_env_vars("thisshouldbevalidtoken")
        assert len(w) == 1
        assert issubclass(w[-1].category, DeprecationWarning)
        assert "deprecated" in str(w[-1].message)

    assert os.environ.get("MLFLOW_TRACKING_URI") == content["mlflow_server"]
    assert os.environ.get("MLFLOW_S3_ENDPOINT_URL") == content["minio_server"]
    assert os.environ.get("AWS_SECRET_ACCESS_KEY") == content["minio_secret_key"]
    assert os.environ.get("AWS_ACCESS_KEY_ID") == content["minio_access_key"]
    assert os.environ.get("MLFLOW_TRACKING_TOKEN") == "thisshouldbevalidtoken"


@responses.activate
def test_get_best_run():
    env_content = {
        "minio_secret_key": "minio_secret",
        "minio_access_key": "minio_access",
        "minio_server": "http://192.198.0.1:9000",
        "mlflow_server": "http://192.198.0.1:5000",
    }
    responses.add(
        responses.Response(
            method="GET",
            url=SERVER_HOST + "/api/env_vars/",
            json=env_content,
            status=200,
        )
    )

    mflux_ai.init("thisshouldbevalidtoken")

    headers = {
        "User-Agent": "mlflow-python-client/1.0.0",
        "Accept-Encoding": "gzip, deflate",
        "Accept": "*/*",
        "Connection": "keep-alive",
        "Authorization": "Bearer thisshouldbevalidtoken",
    }
    content = {
        "run": {
            "info": {
                "run_uuid": "123",
                "experiment_id": "2",
                "user_id": "Iver",
                "status": "FINISHED",
                "start_time": "1577979142226",
                "end_time": "1577979155221",
                "artifact_uri": "s3://mlflow/2/123/artifacts",
                "lifecycle_stage": "active",
                "run_id": "123",
            },
            "data": {
                "metrics": [
                    {
                        "key": "error",
                        "value": 1.06968755342329e-05,
                        "timestamp": "1577979154751",
                        "step": "49",
                    }
                ],
                "params": [
                    {"key": "optimizer_name", "value": "FastGANoisyDiscreteOnePlusOne"}
                ],
                "tags": [
                    {"key": "mlflow.user", "value": "Iver"},
                    {
                        "key": "mlflow.source.name",
                        "value": "C:/Users/Iver/Code/mflux-quickstart/nevergrad_example.py",
                    },
                    {"key": "mlflow.source.type", "value": "LOCAL"},
                ],
            },
        }
    }
    # Mock two different URLs to account for differences in mlflow versions
    responses.add(
        responses.Response(
            method="GET",
            url=env_content["mlflow_server"] + "/api/2.0/preview/mlflow/runs/get",
            json=content,
            status=200,
            headers=headers,
        ),
    )
    responses.add(
        responses.Response(
            method="GET",
            url=env_content["mlflow_server"] + "/api/2.0/mlflow/runs/get",
            json=content,
            status=200,
            headers=headers,
        ),
    )

    headers = {
        "Content-Type": "application/vnd.aiascience.mflux+json; version=0.4",
        "Authorization": "api-key {}".format("thisshouldbevalidtoken"),
    }
    content = {"run_uuid": "123"}
    url = (
        SERVER_HOST
        + "/api/best_run_by_model_group/best_run/?model_group_name={}".format(
            "model_name"
        )
    )
    responses.add(
        responses.Response(
            method="GET", url=url, json=content, status=200, headers=headers
        )
    )

    best_run = mflux_ai.get_best_run("model_name")
    assert isinstance(best_run, mlflow.entities.run.Run)
    assert best_run.info.run_uuid == content["run_uuid"]


tests/test_mflux_ai.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for the `mflux_ai` package."""
import os
import warnings

import responses

import mflux_ai
from mflux_ai.mflux_ai import SERVER_HOST


@responses.activate
def test_mflux_ai_init():
    """Test the init function."""
    content = {
        "minio_secret_key": "minio_secret",
        "minio_access_key": "minio_access",
        "minio_server": "http://192.198.0.1:9000",
        "mlflow_server": "http://192.198.0.1:5000",
    }

    responses.add(
        responses.Response(
            method="GET", url=SERVER_HOST + "/api/env_vars/", json=content, status=200
        )
    )

    mflux_ai.init("thisshouldbevalidtoken")
    assert os.environ.get("MLFLOW_TRACKING_URI") == content["mlflow_server"]
    assert os.environ.get("MLFLOW_S3_ENDPOINT_URL") == content["minio_server"]
    assert os.environ.get("AWS_SECRET_ACCESS_KEY") == content["minio_secret_key"]
    assert os.environ.get("AWS_ACCESS_KEY_ID") == content["minio_access_key"]
    assert os.environ.get("MLFLOW_TRACKING_TOKEN") == "thisshouldbevalidtoken"


@responses.activate
def test_mflux_ai_deprecated_set_env_vars():
    """Test the deprecated set_env_vars function."""
    content = {
        "minio_secret_key": "minio_secret",
        "minio_access_key": "minio_access",
        "minio_server": "http://192.198.0.1:9000",
        "mlflow_server": "http://192.198.0.1:5000",
    }

    responses.add(
        responses.Response(
            method="GET", url=SERVER_HOST + "/api/env_vars/", json=content, status=200
        )
    )

    with warnings.catch_warnings(record=True) as w:
        # Cause all warnings to always be triggered.
        warnings.simplefilter("always")
        mflux_ai.set_env_vars("thisshouldbevalidtoken")
        assert len(w) == 1
        assert issubclass(w[-1].category, DeprecationWarning)
        assert "deprecated" in str(w[-1].message)

    assert os.environ.get("MLFLOW_TRACKING_URI") == content["mlflow_server"]
    assert os.environ.get("MLFLOW_S3_ENDPOINT_URL") == content["minio_server"]
    assert os.environ.get("AWS_SECRET_ACCESS_KEY") == content["minio_secret_key"]
    assert os.environ.get("AWS_ACCESS_KEY_ID") == content["minio_access_key"]
    assert os.environ.get("MLFLOW_TRACKING_TOKEN") == "thisshouldbevalidtoken"


tests/test_moved_mflux_ai.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""Tests for `mflux_ai.mflux.ai`, which is now deprecated."""
import os
import warnings

import responses

from mflux_ai import mflux_ai
from mflux_ai import core as mflux_ai_core


@responses.activate
def test_mflux_ai_init():
    """Test that the init function still works, although it has been moved. It should raise a
     DeprecationWarning."""
    content = {
        "minio_secret_key": "minio_secret",
        "minio_access_key": "minio_access",
        "minio_server": "http://192.198.0.1:9000",
        "mlflow_server": "http://192.198.0.1:5000",
    }

    responses.add(
        responses.Response(
            method="GET", url=mflux_ai.SERVER_HOST + "/api/env_vars/", json=content, status=200
        )
    )

    with warnings.catch_warnings(record=True) as w:
        # Cause all warnings to always be triggered.
        warnings.simplefilter("always")
        mflux_ai.init("thisshouldbevalidtoken")
        assert len(w) == 1
        assert issubclass(w[-1].category, DeprecationWarning)
        assert "has been moved" in str(w[-1].message)

    assert os.environ.get("MLFLOW_TRACKING_URI") == content["mlflow_server"]
    assert os.environ.get("MLFLOW_S3_ENDPOINT_URL") == content["minio_server"]
    assert os.environ.get("AWS_SECRET_ACCESS_KEY") == content["minio_secret_key"]
    assert os.environ.get("AWS_ACCESS_KEY_ID") == content["minio_access_key"]
    assert os.environ.get("MLFLOW_TRACKING_TOKEN") == "thisshouldbevalidtoken"


@responses.activate
def test_change_deprecated_server_host_variable():
    """Test that the deprecated SERVER_HOST variable still works, although it has been moved.
    Changing this variable should raise a DeprecationWarning."""
    mflux_ai.SERVER_HOST = "https://test.mflux.ai"
    content = {
        "minio_secret_key": "minio_secret",
        "minio_access_key": "minio_access",
        "minio_server": "http://192.198.0.1:9000",
        "mlflow_server": "http://192.198.0.1:5000",
    }

    responses.add(
        responses.Response(
            method="GET", url=mflux_ai.SERVER_HOST + "/api/env_vars/", json=content, status=200
        )
    )

    with warnings.catch_warnings(record=True) as w:
        # Cause all warnings to always be triggered.
        warnings.simplefilter("always")
        mflux_ai_core.init("thisshouldbevalidtoken")
        assert len(w) == 1
        assert issubclass(w[-1].category, DeprecationWarning)
        assert "The internal SERVER_HOST variable has been moved" in str(w[-1].message)

    assert os.environ.get("MLFLOW_TRACKING_URI") == content["mlflow_server"]
    assert os.environ.get("MLFLOW_S3_ENDPOINT_URL") == content["minio_server"]
    assert os.environ.get("AWS_SECRET_ACCESS_KEY") == content["minio_secret_key"]
    assert os.environ.get("AWS_ACCESS_KEY_ID") == content["minio_access_key"]
    assert os.environ.get("MLFLOW_TRACKING_TOKEN") == "thisshouldbevalidtoken"


mflux_ai/mflux_ai.py
import warnings

from mflux_ai.core import init as core_init
from mflux_ai.core import get_minio_client as core_get_minio_client
from mflux_ai.core import ensure_bucket_exists as core_ensure_bucket_exists
from mflux_ai.core import put_dataset as core_put_dataset
from mflux_ai.core import get_dataset as core_get_dataset


SERVER_HOST = "https://www.mflux.ai"
_minio_client = None


def init(project_token):
    """
    Note: This function has been moved from mflux_ai.mflux_ai to mflux_ai
    """
    warnings.warn(
        "The init function has been moved. If you are doing \n"
        "from mflux_ai import mflux_ai\n"
        'mflux_ai.init("your_project_token_goes_here")\n'
        "then you should instead do this:\n"
        "import mflux_ai\n"
        'mflux_ai.init("your_project_token_goes_here")',
        DeprecationWarning,
    )
    core_init(project_token)


def get_minio_client():
    """
    Note: This function has been moved from mflux_ai.mflux_ai to mflux_ai
    """
    warnings.warn(
        "The get_minio_client function has been moved. If you are doing \n"
        "from mflux_ai import mflux_ai\n"
        "mflux_ai.get_minio_client(...)\n"
        "then you should instead do this:\n"
        "import mflux_ai\n"
        "mflux_ai.get_minio_client(...)",
        DeprecationWarning,
    )
    return core_get_minio_client()


def ensure_bucket_exists(bucket_name):
    """
    Note: This function has been moved from mflux_ai.mflux_ai to mflux_ai
    """
    warnings.warn(
        "The ensure_bucket_exists function has been moved. If you are doing \n"
        "from mflux_ai import mflux_ai\n"
        "mflux_ai.ensure_bucket_exists(...)\n"
        "then you should instead do this:\n"
        "import mflux_ai\n"
        "mflux_ai.ensure_bucket_exists(...)",
        DeprecationWarning,
    )
    return core_ensure_bucket_exists(bucket_name)


def put_dataset(value, object_name, bucket_name="datasets"):
    """
    Note: This function has been moved from mflux_ai.mflux_ai to mflux_ai
    """
    warnings.warn(
        "The put_dataset function has been moved. If you are doing \n"
        "from mflux_ai import mflux_ai\n"
        "mflux_ai.ensure_bucket_exists(...)\n"
        "then you should instead do this:\n"
        "import mflux_ai\n"
        "mflux_ai.ensure_bucket_exists(...)",
        DeprecationWarning,
    )
    core_put_dataset(value, object_name, bucket_name)


def get_dataset(object_name, bucket_name="datasets"):
    """
    Note: This function has been moved from mflux_ai.mflux_ai to mflux_ai
    """
    warnings.warn(
        "The get_dataset function has been moved. If you are doing \n"
        "from mflux_ai import mflux_ai\n"
        "mflux_ai.ensure_bucket_exists(...)\n"
        "then you should instead do this:\n"
        "import mflux_ai\n"
        "mflux_ai.ensure_bucket_exists(...)",
        DeprecationWarning,
    )
    core_get_dataset(object_name, bucket_name)


mflux_ai/core.py
# -*- coding: utf-8 -*-

import io
import os
import warnings

import joblib
import requests
from minio import Minio
from minio.error import (
    BucketAlreadyOwnedByYou,
    BucketAlreadyExists,
    NoSuchBucket,
    ResponseError,
)
import mlflow

SERVER_HOST = "https://www.mflux.ai"

_minio_client = None


def init(project_token):
    """
    Fetch connection strings and set them as environment variables that MLflow understands and
    uses when it connects to your MFlux.ai server.

    :param project_token: A secret string that is specific to an MFlux.ai project that you have
        access to.
    """
    global _minio_client
    global SERVER_HOST
    if "your_" in project_token.lower():
        print(
            'Warning: "{}" looks like an invalid project token. Go to'
            " {}/dashboard/ to obtain your project token.".format(
                project_token, SERVER_HOST
            )
        )

    from mflux_ai.mflux_ai import SERVER_HOST as DEPRECATED_SERVER_HOST_VARIABLE

    if DEPRECATED_SERVER_HOST_VARIABLE != "https://www.mflux.ai":
        warnings.warn(
            "The internal SERVER_HOST variable has been moved. If you are doing\n"
            "from mflux_ai import mflux_ai\n"
            "mflux_ai.SERVER_HOST = ...\n"
            "or\n"
            "import mflux_ai\n"
            "mflux_ai.mflux_ai.SERVER_HOST = ...\n"
            "then you should instead do this:\n"
            "import mflux_ai\n"
            "mflux_ai.core.SERVER_HOST = ...",
            DeprecationWarning,
        )
        # Note: This will be changed from warning to error in a future version
        SERVER_HOST = DEPRECATED_SERVER_HOST_VARIABLE

    headers = {
        "Accept": "application/vnd.aiascience.mflux+json; version=0.4",
        "Authorization": "api-key {}".format(project_token),
    }
    url = SERVER_HOST + "/api/env_vars/"
    try:
        response = requests.get(url, headers=headers)
    except requests.exceptions.RequestException:
        print(
            "Error: Could not connect to the MFlux.ai server ({}). If this issue persists,"
            " please contact MFlux.ai's support.".format(SERVER_HOST)
        )
        raise

    if response.status_code != 200:
        if response.status_code == 406 and "Invalid version" in str(response.content):
            # We import __version__ here to avoid circular imports
            from . import __version__

            raise Exception(
                "Error: Bad status code {}. This may indicate your mflux-ai python package"
                " needs to be upgraded to a newer version. Currently, mflux-ai=={} is"
                " installed. Go to https://pypi.org/project/mflux-ai/ to see what the latest"
                " version of mflux-ai is.".format(response.status_code, __version__)
            )
        elif response.status_code == 204:
            raise Exception(
                "Error: Bad status code {}. This may indicate that your project token is"
                " invalid.".format(response.status_code)
            )
        else:
            raise Exception(
                "Error: Bad status code {}. If this issue persists,"
                " please contact MFlux.ai's support.".format(response.status_code)
            )

    data = response.json()
    if not data.get("mlflow_server", None):
        print("Error: Could not fetch mlflow_server connection string")
    if not data.get("minio_server", None):
        print("Error: Could not fetch minio_server connection string")
    if not data.get("minio_access_key", None):
        print("Error: Could not fetch minio_access_key connection string")
    if not data.get("minio_secret_key", None):
        print("Error: Could not fetch minio_secret_key connection string")

    _minio_client = None
    os.environ["MLFLOW_TRACKING_URI"] = data.get("mlflow_server", None)
    os.environ["MLFLOW_S3_ENDPOINT_URL"] = data.get("minio_server", None)
    os.environ["AWS_ACCESS_KEY_ID"] = data.get("minio_access_key", None)
    os.environ["AWS_SECRET_ACCESS_KEY"] = data.get("minio_secret_key", None)
    os.environ["MLFLOW_TRACKING_TOKEN"] = project_token


def get_minio_client():
    """Return a Minio instance. If one has been created earlier, return the same instance."""
    global _minio_client

    assert os.environ.get("MLFLOW_S3_ENDPOINT_URL", None) is not None
    assert os.environ.get("AWS_ACCESS_KEY_ID", None) is not None
    assert os.environ.get("AWS_SECRET_ACCESS_KEY", None) is not None
    if not _minio_client:
        is_secure = "https://" in os.environ["MLFLOW_S3_ENDPOINT_URL"]
        endpoint = (
            os.environ["MLFLOW_S3_ENDPOINT_URL"]
            .replace("http://", "")
            .replace("https://", "")
            .replace(":9000/", ":9000")  # TODO: Remove this replace
        )
        _minio_client = Minio(
            endpoint,
            access_key=os.environ["AWS_ACCESS_KEY_ID"],
            secret_key=os.environ["AWS_SECRET_ACCESS_KEY"],
            secure=is_secure,
        )

    return _minio_client


def ensure_bucket_exists(bucket_name):
    """Create a bucket with the given bucket_name if it doesn't already exist."""
    minio_client = get_minio_client()

    try:
        minio_client.make_bucket(bucket_name)
    except BucketAlreadyOwnedByYou as err:
        pass
    except BucketAlreadyExists as err:
        pass
    except ResponseError as err:
        raise


def put_dataset(value, object_name, bucket_name="datasets"):
    """
    Store an object/dataset in the MFlux.ai cloud. It gets pickled by joblib.dump.

    :param value: The object/dataset to store. It should be picklable.
    :param object_name: The name of the dataset
    :param bucket_name: (Optional, defaults to "datasets") Name of the bucket to store the
        object/dataset in. Think of it as a folder. This name must not contain underscores. For
        more info on bucket name restrictions, see
        https://docs.aws.amazon.com/AmazonS3/latest/dev/BucketRestrictions.html
        If the bucket does not already exist, it will be created.
    :return:
    """
    minio_client = get_minio_client()

    in_memory_file = io.BytesIO()
    joblib.dump(value, in_memory_file, compress=True)
    num_bytes = in_memory_file.getbuffer().nbytes

    # Prepare for the read() call on the in-memory file object
    in_memory_file.seek(0)

    for _ in range(2):
        try:
            minio_client.put_object(bucket_name, object_name, in_memory_file, num_bytes)
            break  # Success! Now exit the loop
        except ResponseError as err:
            print(err)
            break  # Failure, no retry
        except NoSuchBucket:
            ensure_bucket_exists(bucket_name)
            continue  # Try the object upload again now that the bucket has been created


def get_dataset(object_name, bucket_name="datasets"):
    """
    Retrieve the object/dataset with the given object_name from the MFlux.ai cloud. The object
    gets unpickled by joblib.load.
    """
    minio_client = get_minio_client()

    in_memory_file = io.BytesIO()
    try:
        data = minio_client.get_object(bucket_name, object_name)
        for d in data.stream(32 * 1024):
            in_memory_file.write(d)
    except ResponseError as err:
        raise

    # Prepare for the read() call on the in-memory file object
    in_memory_file.seek(0)

    return joblib.load(in_memory_file)


def get_best_run(model_group_name):
    """
    Fetch the run id for the best run in a model group

    :param model_group_name:
    """

    if "MLFLOW_TRACKING_TOKEN" not in os.environ:
        raise Exception(
            "Could not find the MFlux.ai project token for authentication. Please call mflux_ai.init('your_project_token_goes_here') before calling mflux_ai.get_best_run()"
        )

    project_token = os.environ["MLFLOW_TRACKING_TOKEN"]

    headers = {
        "Content-Type": "application/vnd.aiascience.mflux+json; version=0.4",
        "Authorization": "api-key {}".format(project_token),
    }
    url = (
        SERVER_HOST
        + "/api/best_run_by_model_group/best_run/?model_group_name={}".format(
            model_group_name
        )
    )
    try:
        response = requests.get(url, headers=headers)
    except requests.exceptions.RequestException:
        print(
            "Error: Could not connect to the MFlux.ai server ({}). If this issue persists,"
            " please contact MFlux.ai's support.".format(SERVER_HOST)
        )
        raise

    if response.status_code != 200:
        if response.status_code == 406 and "Invalid version" in str(response.content):
            # We import __version__ here to avoid circular imports
            from . import __version__

            raise Exception(
                "Error: Bad status code {}. This may indicate your mflux-ai python package"
                " needs to be upgraded to a newer version. Currently, mflux-ai=={} is"
                " installed. Go to https://pypi.org/project/mflux-ai/ to see what the latest"
                " version of mflux-ai is.".format(response.status_code, __version__)
            )
        elif response.status_code == 404:
            raise Exception(
                "Error: Bad status code {}. This may indicate that the model group does"
                " not exists.".format(response.status_code)
            )
        else:
            raise Exception(
                "Error: Bad status code {}. If this issue persists,"
                " please contact MFlux.ai's support.".format(response.status_code)
            )

    data = response.json()
    run_id = data["run_uuid"]

    if run_id is None:
        return None

    client = mlflow.tracking.MlflowClient()
    run = client.get_run(run_id)
    return run
